In [1]:
import os
print(os.getcwd())

d:\MLOps\customer-churn\experiments


In [2]:
os.chdir("../")
print(os.getcwd())

d:\MLOps\customer-churn


In [3]:
import pandas as pd
import numpy as np
df=pd.read_csv("artifacts\dataset.csv")

In [4]:
df

CustomerID   Age  Gender  Tenure  Usage Frequency  Support Calls  \
0              2.0  30.0  Female    39.0             14.0            5.0   
1              3.0  65.0  Female    49.0              1.0           10.0   
2              4.0  55.0  Female    14.0              4.0            6.0   
3              5.0  58.0    Male    38.0             21.0            7.0   
4              6.0  23.0    Male    32.0             20.0            5.0   
...            ...   ...     ...     ...              ...            ...   
440828    449995.0  42.0    Male    54.0             15.0            1.0   
440829    449996.0  25.0  Female     8.0             13.0            1.0   
440830    449997.0  26.0    Male    35.0             27.0            1.0   
440831    449998.0  28.0    Male    55.0             14.0            2.0   
440832    449999.0  31.0    Male    48.0             20.0            1.0   

        Payment Delay Subscription Type Contract Length  Total Spend  \
0                18.0          Standard          Annual       932.00   
1                 8.0             Basic         Monthly       557.00   
2                18.0             Basic       Quarterly       185.00   
3                 7.0          Standard         Monthly       396.00   
4                 8.0             Basic         Monthly       617.00   
...               ...               ...             ...          ...   
440828            3.0           Premium          Annual       716.38   
440829           20.0           Premium          Annual       745.38   
440830            5.0          Standard       Quarterly       977.31   
440831            0.0          Standard       Quarterly       602.55   
440832           14.0           Premium       Quarterly       567.77   

        Last Interaction  Churn  
0                   17.0    1.0  
1                    6.0    1.0  
2                    3.0    1.0  
3                   29.0    1.0  
4                   20.0    1.0  
...                  ...    ...  
440828               8.0    0.0  
440829               2.0    0.0  
440830               9.0    0.0  
440831               2.0    0.0  
440832              21.0    0.0  

[440833 rows x 12 columns]

In [5]:
df['Contract Length'].value_counts()

Contract Length
Annual       177198
Quarterly    176530
Monthly       87104
Name: count, dtype: int64

In [6]:
df[df['Churn']==1]['Contract Length'].value_counts()

Contract Length
Monthly      87104
Annual       81646
Quarterly    81249
Name: count, dtype: int64

In [7]:
df[df['Churn']==0]['Contract Length'].value_counts()

Contract Length
Annual       95552
Quarterly    95281
Name: count, dtype: int64

Conclusion: We can clearly see that monthly customers are churners.

In [8]:
df[df['Contract Length']=='Monthly']

CustomerID   Age  Gender  Tenure  Usage Frequency  Support Calls  \
1              3.0  65.0  Female    49.0              1.0           10.0   
3              5.0  58.0    Male    38.0             21.0            7.0   
4              6.0  23.0    Male    32.0             20.0            5.0   
11            14.0  52.0  Female    21.0              6.0            3.0   
18            21.0  24.0    Male    44.0             13.0            5.0   
...            ...   ...     ...     ...              ...            ...   
253683    260797.0  64.0  Female    19.0             28.0            6.0   
253684    260798.0  28.0    Male    50.0             15.0            2.0   
253687    260801.0  22.0    Male    10.0             17.0            5.0   
253690    260804.0  24.0    Male    54.0             21.0            7.0   
253692    260806.0  29.0  Female    46.0              9.0            8.0   

        Payment Delay Subscription Type Contract Length  Total Spend  \
1                 8.0             Basic         Monthly       557.00   
3                 7.0          Standard         Monthly       396.00   
4                 8.0             Basic         Monthly       617.00   
11               26.0           Premium         Monthly       830.00   
18                4.0           Premium         Monthly       669.00   
...               ...               ...             ...          ...   
253683           21.0           Premium         Monthly       810.08   
253684           13.0           Premium         Monthly       427.31   
253687           14.0          Standard         Monthly       494.73   
253690            8.0           Premium         Monthly       102.56   
253692           19.0             Basic         Monthly       248.28   

        Last Interaction  Churn  
1                    6.0    1.0  
3                   29.0    1.0  
4                   20.0    1.0  
11                  19.0    1.0  
18                  13.0    1.0  
...                  ...    ...  
253683              20.0    1.0  
253684               5.0    1.0  
253687              27.0    1.0  
253690               2.0    1.0  
253692              21.0    1.0  

[87104 rows x 12 columns]

Conclusion: contract length is monthly but their tenure is not 1??? maybe the data is wrong

In [11]:
inconsistent_rows = df[(df['Contract Length'] == 'Monthly') & (df['Tenure'] != 1)]
inconsistent_rows

CustomerID   Age  Gender  Tenure  Usage Frequency  Support Calls  \
1              3.0  65.0  Female    49.0              1.0           10.0   
3              5.0  58.0    Male    38.0             21.0            7.0   
4              6.0  23.0    Male    32.0             20.0            5.0   
11            14.0  52.0  Female    21.0              6.0            3.0   
18            21.0  24.0    Male    44.0             13.0            5.0   
...            ...   ...     ...     ...              ...            ...   
253683    260797.0  64.0  Female    19.0             28.0            6.0   
253684    260798.0  28.0    Male    50.0             15.0            2.0   
253687    260801.0  22.0    Male    10.0             17.0            5.0   
253690    260804.0  24.0    Male    54.0             21.0            7.0   
253692    260806.0  29.0  Female    46.0              9.0            8.0   

        Payment Delay Subscription Type Contract Length  Total Spend  \
1                 8.0             Basic         Monthly       557.00   
3                 7.0          Standard         Monthly       396.00   
4                 8.0             Basic         Monthly       617.00   
11               26.0           Premium         Monthly       830.00   
18                4.0           Premium         Monthly       669.00   
...               ...               ...             ...          ...   
253683           21.0           Premium         Monthly       810.08   
253684           13.0           Premium         Monthly       427.31   
253687           14.0          Standard         Monthly       494.73   
253690            8.0           Premium         Monthly       102.56   
253692           19.0             Basic         Monthly       248.28   

        Last Interaction  Churn  
1                    6.0    1.0  
3                   29.0    1.0  
4                   20.0    1.0  
11                  19.0    1.0  
18                  13.0    1.0  
...                  ...    ...  
253683              20.0    1.0  
253684               5.0    1.0  
253687              27.0    1.0  
253690               2.0    1.0  
253692              21.0    1.0  

[85615 rows x 12 columns]

In [12]:
consistent_rows = df[(df['Contract Length'] == 'Monthly') & (df['Tenure'] == 1)]
consistent_rows

CustomerID   Age  Gender  Tenure  Usage Frequency  Support Calls  \
227          236.0  33.0  Female     1.0              1.0            7.0   
492          501.0  29.0  Female     1.0             16.0           10.0   
638          647.0  19.0  Female     1.0             14.0            9.0   
783          792.0  22.0  Female     1.0             30.0            6.0   
871          880.0  46.0    Male     1.0             21.0            6.0   
...            ...   ...     ...     ...              ...            ...   
252808    259922.0  49.0    Male     1.0              4.0            0.0   
253209    260323.0  18.0  Female     1.0              5.0            0.0   
253313    260427.0  42.0    Male     1.0              2.0            1.0   
253375    260489.0  28.0  Female     1.0             23.0            5.0   
253377    260491.0  48.0  Female     1.0             24.0            4.0   

        Payment Delay Subscription Type Contract Length  Total Spend  \
227              18.0             Basic         Monthly       411.00   
492              25.0           Premium         Monthly       464.00   
638               6.0           Premium         Monthly       356.00   
783               0.0             Basic         Monthly       639.00   
871              12.0             Basic         Monthly       550.00   
...               ...               ...             ...          ...   
252808           19.0          Standard         Monthly       350.27   
253209           12.0          Standard         Monthly       235.25   
253313            1.0             Basic         Monthly       832.29   
253375            9.0           Premium         Monthly       665.01   
253377           17.0          Standard         Monthly       478.51   

        Last Interaction  Churn  
227                 16.0    1.0  
492                  1.0    1.0  
638                 14.0    1.0  
783                 23.0    1.0  
871                 28.0    1.0  
...                  ...    ...  
252808              22.0    1.0  
253209              17.0    1.0  
253313              29.0    1.0  
253375              23.0    1.0  
253377              10.0    1.0  

[1489 rows x 12 columns]

Conclusion: Out of 87k only 1489 samples are valid tenure=1 and contract length is monthly

In [13]:
df.shape

(440833, 12)

In [15]:
df['CustomerID'].nunique()

440832

Conclusion: Customer and samples are same so no chance of downgrade